rfc-174-consolidate-frontend-rendering-apps.md

```diff
@@ -0,0 +1,330 @@
+
+#### Making changes to the layout is hard
+Having more than one application makes changing the layout far more complicated than necessary.
+
+##### Case study
+The action required in a [recent incident] was to remove an email form field from the feedback component in the footer of all pages and replace it with a link to a survey. This fix should have involved two simple steps: remove the offending form field, and get the URL of the current page and include it in a new link. This should have been easy, but took three lead developers and two 2nd line developers five hours to implement.
+
+This was because of the time it took to get the frontend stack running locally to be able to test the changes. It required a local application pointing at a local version of static, each using a local version of `govuk_publishing_components`. The feedback form is part of `govuk_publishing_components, but part of the layout called from static.
+
+This was also because the feedback form included custom JavaScript to get the current page URL. This was not clear to those handling the incident and caused a delay in solving it. With a single application, the current page URL could be handled internally with standard Ruby helpers, making the implementation and solving the problem far simpler.
+
+### Translations
+There is existing tech debt for [multiple frontend applications use the same translated phrases]. These are some of the existing issues that will be addressed by having a single rendering app:
+
+1. Different pages on the site can have different translations for the same English word
+2. Some translations will be missing on pages rendered by different frontend applications
+3. Difficult to get visibility on which phrases are translated and which ones aren't
+
+## Proposal
+
+We propose merging the apps into a single rendering app. This will solve some of the issues in this RFC and allow us to solve others in the future - for example, removing `static`.
+
+We propose merging all apps into an existing app. The benefit of this is that even if we aren’t able to consolidate all of the applications, we would still reduce the maintenance burden for each app merged. We also won’t have created anything new, and won’t have to manage any extra conflicting patterns.
+
+We can steadily move routes, controllers and models from the candidate (the app being merged) into the host (the app we are merging into) until the host is serving all the routes.
+
+We propose using [frontend] as the host app. This is because:
+
+* `frontend` contains some of the more complicated document types like licence_transaction and local_transaction, that have multiple pages and interactive elements, which would be harder to consolidate into another app
+* `frontend` has a one-controller per document type layout, which is preferable for document types with multiple pages as it follows [rails conventions for controllers] to do the orchestration.
```

Comment:
The controller-per-document-type approach does help in a number of ways. In order to do this though, (somewhat unusually for GOV.UK apps) we use [a routing constraint based on document type](https://github.com/alphagov/frontend/blob/af88146e66d7693085e44a92fb4281c25ecf1400/config/routes.rb#L60). This retrieves the content item from content store as part of the routing of a request and checks the document type to determine which controller should handle it.

The content item is cached in the request object, so should only be retrieved once - even though it may be used multiple times in route resolution. Additionally, the controller for document types that use content items also [retrieves the content item from the request object](https://github.com/alphagov/frontend/blob/main/app/controllers/content_items_controller.rb#L25).

I don't think this is a problem, and might well be a useful approach when introducing new types to the app, however it is probably not well known so should probably be socialised at least. It would probably be a Bad Thing:tm: if our single frontend app made duplicate requests to our APIs!

Reply:
Oh that's interesting point. Thanks Simon! We were concerned that we'd be retrieving the content-item twice. But you're right, getting the content item from the request object rather than an api call is a slightly odd pattern.

Reply:
Oh, that's a shame, and also good to know. 😬